src/JustSaying/Fluent/PublishConfig/DynamicAddressMessagePublisher.cs

```diff
@@ -0,0 +1,112 @@
+using System.Collections.Concurrent;
+using JustSaying.Messaging;
+using JustSaying.Messaging.Interrogation;
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Fluent;
+
+internal sealed class DynamicAddressMessagePublisher(
+    string topicArnTemplate,
+    Func<string, Message, string> topicAddressCustomizer,
+    Func<string, StaticAddressPublicationConfiguration> staticConfigBuilder,
+    ILoggerFactory loggerFactory) : IMessagePublisher, IMessageBatchPublisher
+{
+    private readonly string _topicArnTemplate = topicArnTemplate;
+    private readonly ConcurrentDictionary<string, IMessagePublisher> _publisherCache = new();
+    private readonly ConcurrentDictionary<string, IMessageBatchPublisher> _batchPublisherCache = new();
+    private readonly ConcurrentDictionary<string, SemaphoreSlim> _topicCreationLocks = new();
+    private readonly ILogger<DynamicMessagePublisher> _logger = loggerFactory.CreateLogger<DynamicMessagePublisher>();
+    private readonly Func<string, Message, string> _topicAddressCustomizer = topicAddressCustomizer;
+    private readonly Func<string, StaticAddressPublicationConfiguration> _staticConfigBuilder = staticConfigBuilder;
+
+    /// <inheritdoc/>
+    public InterrogationResult Interrogate()
+    {
+        var publishers = _publisherCache.Keys.OrderBy(x => x).ToDictionary(x => x, x => _publisherCache[x].Interrogate());
+        var batchPublishers = _batchPublisherCache.Keys.OrderBy(x => x).ToDictionary(x => x, x => _batchPublisherCache[x].Interrogate());
+
+        return new InterrogationResult(new
+        {
+            Publishers = publishers,
+            BatchPublishers = batchPublishers,
+        });
+    }
+
+    /// <inheritdoc/>
+    public Task StartAsync(CancellationToken stoppingToken) => Task.CompletedTask;
+
+    /// <inheritdoc/>
+    public async Task PublishAsync(Message message, PublishMetadata metadata, CancellationToken cancellationToken)
+    {
+        string topicArn = _topicAddressCustomizer(_topicArnTemplate, message);
+        if (_publisherCache.TryGetValue(topicArn, out var publisher))
+        {
+            await publisher.PublishAsync(message, metadata, cancellationToken).ConfigureAwait(false);
+            return;
+        }
+
+        var lockObj = _topicCreationLocks.GetOrAdd(topicArn, _ => new SemaphoreSlim(1, 1));
+
+        _logger.LogDebug("Publisher for topic {TopicArn} not found, waiting on setup lock", topicArn);
+        await lockObj.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Won't this wait forever and then throw an exception when the timeout expires after the first call as we don't use `Release`?

Reply:
Good catch, we have the same issue in DynamicMessagePublisher